File: 2022/2022W18/PreppindataW18.py
import pandas as pd
import re

final = pd.read_csv(r"C:\Data\PreppinData\2022W18 Input.csv")
final = final.melt(id_vars='Region', var_name='Cols', value_name='Amount')
final['Bike Type'] = final['Cols'].apply(lambda x: re.search('^([^_]+)(?=_)', x).group(1))
final['Measure'] = final['Cols'].apply(lambda x: re.search('(?<=_)([^_]+)$', x).group(1))
final['Month'] = final['Cols'].apply(lambda x: pd.to_datetime(re.sub('.*([A-Za-z]{3})_(\d+).*', '20\\2-\\1-01', x), 
                                                              format='%Y-%b-%d').date())
final = final.drop('Cols', axis=1)
final = final.pivot_table(index=['Bike Type', 'Region', 'Month'], columns='Measure', values='Amount', aggfunc=sum).reset_index()
final = final[['Bike Type', 'Region', 'Month', 'Sales', 'Profit']]

File: 2022/2022W18/preppindataW18.R
library(readr)
library(dplyr)
library(tidyr)

Sys.setlocale("LC_ALL","English")

final <- read_csv("C:/Data/PreppinData/2022W18 Input.csv") %>%
  pivot_longer(cols = -'Region', names_to = c('Bike Type', 'Mth', 'Yr', 'Measure'), 
               names_pattern='(\\w+)___(\\w+)_(\\d+)+___(\\w+)', values_to = 'Amount') %>%
  mutate('Month' = as.Date(paste0('20',`Yr`,'-',`Mth`,'-01'), format='%Y-%b-%d')) %>%
  pivot_wider(id_cols = -'Measure', names_from='Measure', values_from='Amount', values_fn=sum) %>%
  select(c('Bike Type', 'Region', 'Month', 'Sales', 'Profit'))

View(final)
